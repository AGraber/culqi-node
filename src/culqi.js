import request from 'request'

const baseUrl_integracion = 'https://integ-pago.culqi.com/api/v1';
const baseUrl_produccion = 'https://pago.culqi.com/api/v1';

const paths = {
  crearToken: '/tokens',
  crearCargo: '/cargos'
}

const _createPromise = (url, method, headers, body, validateParams) => {
  return new Promise((resolve, reject) => {

    if (validateParams && body) {
      const keys = Object.keys(body); 

      //console.log('body', body);   

      for (let i in validateParams) {
        if (keys.indexOf(validateParams[i]) == -1) {
          return reject({
            body: {
              objeto: 'error',
              mensaje: 'Falta el parÃ¡metro: ' + validateParams[i]
            }
          });
        }
      }
    }
    request({ url, method, headers, json: true, body }, (error, response, body) => {
      console.log('----------------------------');
      if (!error && response.statusCode == 200) {
        return resolve(body);
      }
      return reject(response);
    })
  })
}

class Culqi {

  constructor (codigo_comercio, llave_comercio, env) {
    this.codigo_comercio = codigo_comercio;
    this.llave_comercio = llave_comercio;
    this.baseUrl = baseUrl_integracion;

    // Validate env
    if (env && (env.toLowerCase() == 'prod' || env.toLowerCase() == 'produccion')) {
      this.baseUrl = baseUrl_produccion;
    }

    // Set bearer header and others
    this._headers = {
      'Authorization': 'Bearer ' + this.llave_comercio,
      'User-Agent': 'request',
      'Accept': 'application/json'
    }
  }

  crearToken (params) {
    console.log('-------- crearToken --------');
    console.log('url: ', this.baseUrl + paths.crearToken);
    console.log('params: ', params);
    
    const url = this.baseUrl + paths.crearToken;
    const fields = ["correo_electronico", "nombre", "apellido", "numero", "cvv", "m_exp", "a_exp", "guardar"];

    return _createPromise(url, 'POST', this._headers, params, fields);
  }

  crearCargo (params) {
    console.log('-------- crearCargo --------');
    console.log('url: ', this.baseUrl + paths.crearCargo);
    console.log('params: ', params);
    
    const url = this.baseUrl + paths.crearCargo;
    const fields = ["token", "moneda", "monto", "descripcion", "pedido", "codigo_pais", "ciudad", "usuario", "direccion", "telefono", "nombres", "apellidos", "correo_electronico"];

    return _createPromise(url, 'POST', this._headers, params, fields);
  }
}



//Compat mode
module.exports = Culqi;
//export default Culqi;