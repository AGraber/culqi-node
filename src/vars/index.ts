export default {
  baseEndpoint: {
    domain: 'api.culqi.com',
    basePath: '/v2',
  },
  baseSecureEndpoint: {
    domain: 'secure.culqi.com',
    basePath: '/v2',
  },
  basePaths: {
    tokens: '/tokens',
    charges: '/charges',
    refunds: '/refunds',
    plans: '/plans',
    subscriptions: '/subscriptions',
    customers: '/customers',
    cards: '/cards',
    events: '/events',
  },
  publicKey: '',
  privateKey: '',
};
