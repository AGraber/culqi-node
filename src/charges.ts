import {get, post, patch, HttpRequestOptions} from './request';
import vars from './vars';

export type Charge = {
  object: string;
  id: string;
  creation_date: number;
  amount: number;
  amount_refunded: number;
  current_amount: number;
  installments: number;
  installments_amount: number;
  currency: string;
  email: string;
  description: string;
  source: {
    object: string;
    id: string;
    type: string;
    email: string;
    creation_date: number;
    card_number: string;
    last_four: string;
    active: boolean;
    iin: {
      object: string;
      bin: string;
      card_brand: string;
      card_type: string;
      card_category: string;
      issuer: {
        name: string;
        country: string;
        country_code: string;
        website: string;
        phone_number: string;
      };
      installments_allowed: number[];
    };
    client: {
      ip: string;
      ip_country: string;
      ip_country_code: string;
      browser: string;
      device_fingerprint: string;
      device_type: string;
    };
    outcome: {
      type: string;
      code: string;
      merchant_message: string;
      user_message: string;
    };
  };
  fraud_score: number;
  antifraud_details: {
    first_name: string;
    last_name: string;
    address: string;
    address_city: string;
    country_code: string;
    phone: string;
    object: string;
  };
  dispute: boolean;
  capture: boolean;
  reference_code: string;
  metadata: {[key: string]: string};
  total_fee: number;
  fee_details: {
    fixed_fee: {};
    variable_fee: {
      currency_code: string;
      commision: number;
      total: number;
    };
  };
  paid: boolean;
  statement_descriptor: string;
  total_fee_taxes: number;
  transfer_amount: number;
  duplicated: boolean;
};

type CreateChargeRequest = {
  amount: string;
  currency_code: string;
  email: string;
  source_id: string;
  capture?: boolean;
  description?: string;
  installments?: number;
  metadata?: {[key: string]: string};
  antifraud_details?: {
    first_name: string;
    last_name: string;
    address: string;
    address_city: string;
    country_code: string;
    phone: string;
  };
};

type GetChargeRequest = {
  id: string;
};

type GetChargesRequest = {
  amount?: string;
  min_amount?: string;
  max_amount?: string;
  installments?: string;
  min_installments?: string;
  max_installments?: string;
  currency_code?: string;
  code?: string;
  decline_code?: string;
  fraud_score?: string;
  min_fraud_score?: string;
  max_fraud_score?: string;
  first_name?: string;
  last_name?: string;
  email?: string;
  address?: string;
  address_city?: string;
  phone_number?: string;
  country_code?: string;
  dispute?: string;
  captured?: string;
  duplicated?: string;
  paid?: string;
  customer_id?: string;
  reference?: string;
  creation_date?: string;
  creation_date_from?: string;
  creation_date_to?: string;
  fee?: string;
  min_fee?: string;
  max_fee?: string;
  card_brand?: string;
  card_type?: string;
  device_type?: string;
  bin?: string;
  limit?: string;
  before?: string;
  after?: string;
};

type GetChargesResponse = {
  data: Charge[];
  paging: {
    previous: string;
    next: string;
    cursors: {
      before: string;
      after: string;
    };
    remaining_items: number;
  };
};

type UpdateChargeRequest = {
  id: string;
  metadata?: {[key: string]: string};
};

type CaptureChargeRequest = {
  id: string;
};

export const charges = {
  createCharge: (
    req: CreateChargeRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) => post<Charge>(vars.basePaths.charges, req, extraHttpOptions),
  getCharge: (
    req: GetChargeRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<Charge>(
      `${vars.basePaths.charges}/${req.id}`,
      undefined,
      extraHttpOptions
    ),
  getCharges: (
    req?: GetChargesRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<GetChargesResponse>(
      vars.basePaths.charges,
      req as {[key: string]: string},
      extraHttpOptions
    ),
  updateCharge: (
    req: UpdateChargeRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    patch<Charge>(`${vars.basePaths.charges}/${req.id}`, req, extraHttpOptions),
  captureCharge: (
    req: CaptureChargeRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    post<Charge>(
      `${vars.basePaths.charges}/${req.id}/capture`,
      req,
      extraHttpOptions
    ),
};
