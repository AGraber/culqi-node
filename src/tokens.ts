import {get, post, patch, HttpRequestOptions} from './request';
import vars from './vars';

export type Token = {
  object: 'token';
  id: string;
  type: string;
  email: string;
  creation_date: number;
  card_number: string;
  last_four: string;
  active: boolean;
  iin: {
    object: 'iin';
    bin: string;
    card_brand: string;
    card_type: string;
    card_category: string;
    issuer: {
      name: string;
      country: string;
      country_code: string;
      website: string;
      phone_number: string;
    };
    installments_allowed: number[];
  };
  client: {
    ip: string;
    ip_country: string;
    ip_country_code: string;
    browser: string;
    device_fingerprint: string;
    device_type: string;
  };
  metadata: {[key: string]: string};
};

type CreateTokenRequest = {
  card_number: string;
  cvv: string;
  expiration_month: string;
  expiration_year: string;
  email: string;
  metadata?: {[key: string]: string};
};

type GetTokenRequest = {
  id: string;
};

type GetTokensRequest = {
  creation_date?: string;
  creation_date_to?: string;
  card_brand?: string;
  card_type?: string;
  device_type?: string;
  bin?: string;
  country_code?: string;
  limit?: string;
  before?: string;
  after?: string;
};

type GetTokensResponse = {
  data: Token[];
  paging: {
    previous: string;
    next: string;
    cursors: {
      before: string;
      after: string;
    };
    remaining_items: number;
  };
};

type UpdateTokenRequest = {
  id: string;
  metadata?: {[key: string]: string};
};

export const tokens = {
  createToken: (
    req: CreateTokenRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    post<Token>(vars.basePaths.tokens, req, {
      ...extraHttpOptions,
      useSecureEndpoint: true,
    }),
  getTokens: (
    req?: GetTokensRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<GetTokensResponse>(
      vars.basePaths.tokens,
      req as {[key: string]: string},
      extraHttpOptions
    ),
  getToken: (
    req: GetTokenRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<Token>(
      `${vars.basePaths.tokens}/${req.id}`,
      undefined,
      extraHttpOptions
    ),
  updateToken: (
    req: UpdateTokenRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    patch<Token>(`${vars.basePaths.tokens}/${req.id}`, req, extraHttpOptions),
};
