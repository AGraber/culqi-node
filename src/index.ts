import vars from './vars';
import {tokens} from './tokens';
import {charges} from './charges';
import {refunds} from './refunds';
import {customers} from './customers';

type CulqiOptions = {
  privateKey?: string;
  pciCompliant?: boolean;
  publicKey?: string;
};

export = class Culqi {
  public tokens = tokens;
  public charges = charges;
  public refunds = refunds;
  public customers = customers;

  constructor(options: CulqiOptions = {}) {
    if (!options.privateKey) throw new Error("Provide 'privateKey' property");
    if (options.pciCompliant && !options.publicKey)
      throw new Error("Provide 'publicKey' property");
    if (!options.pciCompliant && options.publicKey)
      throw new Error(
        "Provide 'pciCompliant' options as true. " +
          'It means you acknowledge that you are PCI-compliant in order to execute certain operations.'
      );
    vars.privateKey = options.privateKey;
    if (options.publicKey) vars.publicKey = options.publicKey;
  }
};
