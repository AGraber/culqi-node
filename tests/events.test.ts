import {events} from '../src/events';
import vars from '../src/vars';
import {httpMockFactory} from './request/__mocks__';
import {RequestOptions} from 'https';

describe('events', () => {
  let publicKey: string;
  let privateKey: string;

  // let eventId: string;

  beforeEach(() => {
    privateKey = process.env.CULQI_PRIVATE_KEY || '';
    publicKey = process.env.CULQI_PUBLIC_KEY || '';
    vars.privateKey = privateKey;
    vars.publicKey = publicKey;
    jest.clearAllMocks();
  });

  describe('getEvents', () => {
    it('should change path', () => {
      const mockedHttps = httpMockFactory();
      events.getEvents(undefined, {
        _httpProvider: mockedHttps,
      });
      const c = mockedHttps.request.mock.calls[0][0] as RequestOptions;
      expect(c.path).toMatchSnapshot();
    });

    /*it.only('should get events', async () => {
      const resp = await events.getEvents();
      console.log('resp', resp);
      // expect(resp.data.length).toBeGreaterThan(0);
    });*/
  });

  describe('getEvent', () => {
    it('should change path', () => {
      const mockedHttps = httpMockFactory();
      events.getEvent(
        {
          id: 'sample-event-id',
        },
        {
          _httpProvider: mockedHttps,
        }
      );
      const c = mockedHttps.request.mock.calls[0][0] as RequestOptions;
      expect(c.path).toMatch(
        `${vars.baseEndpoint.basePath}${vars.basePaths.events}/sample-event-id`
      );
    });

    /*it('should get event', async () => {
      const resp = await events.getEvent({
        id: eventId,
      });
      expect(resp.object).toMatchSnapshot();
      expect(resp.type).toMatchSnapshot();
    });*/
  });
});
