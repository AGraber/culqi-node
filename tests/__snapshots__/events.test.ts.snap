// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`events getEvents should change path 1`] = `"/v2/events"`;
